# Goal of this dataset is to predict the best playing position for a player based on his/her stats

# Data Preprocessing

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('/content/players_fifa23.csv')

In [3]:
data.head()

ID            Name            FullName  Age  Height  Weight  \
0  158023        L. Messi        Lionel Messi   35     169      67   
1  165153      K. Benzema       Karim Benzema   34     185      81   
2  188545  R. Lewandowski  Robert Lewandowski   33     185      81   
3  192985    K. De Bruyne     Kevin De Bruyne   31     181      70   
4  231747       K. Mbappé       Kylian Mbappé   23     182      73   

                                           PhotoUrl Nationality  Overall  \
0  https://cdn.sofifa.net/players/158/023/23_60.png   Argentina       91   
1  https://cdn.sofifa.net/players/165/153/23_60.png      France       91   
2  https://cdn.sofifa.net/players/188/545/23_60.png      Poland       91   
3  https://cdn.sofifa.net/players/192/985/23_60.png     Belgium       91   
4  https://cdn.sofifa.net/players/231/747/23_60.png      France       91   

   Potential  ...  LMRating  CMRating  RMRating LWBRating CDMRating RWBRating  \
0         91  ...        91        88        91        67        66        67   
1         91  ...        89        84        89        67        67        67   
2         91  ...        86        83        86        67        69        67   
3         91  ...        91        91        91        82        82        82   
4         95  ...        92        84        92        70        66        70   

   LBRating  CBRating  RBRating GKRating  
0        62        53        62       22  
1        63        58        63       21  
2        64        63        64       22  
3        78        72        78       24  
4        66        57        66       21  

[5 rows x 90 columns]

In [4]:
needed_columns = ['ID', 'Name', 'Age', 'Height', 'Weight',
       'Overall', 'Potential', 'Growth', 'TotalStats',
       'BaseStats', 'BestPosition', 'Club', 'ValueEUR', 'WageEUR',
       'ReleaseClause', 'ContractUntil', 'ClubJoined', 'OnLoad',
       'PreferredFoot', 'IntReputation', 'WeakFoot','Nationality',
       'SkillMoves', 'AttackingWorkRate', 'DefensiveWorkRate', 'PaceTotal',
       'ShootingTotal', 'PassingTotal', 'DribblingTotal', 'DefendingTotal',
       'PhysicalityTotal', 'Crossing', 'Finishing', 'HeadingAccuracy',
       'ShortPassing', 'Volleys', 'Dribbling', 'Curve', 'FKAccuracy',
       'LongPassing', 'BallControl', 'Acceleration', 'SprintSpeed', 'Agility',
       'Reactions', 'Balance', 'ShotPower', 'Jumping', 'Stamina', 'Strength',
       'LongShots', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Marking', 'StandingTackle', 'SlidingTackle',
       'GKDiving', 'GKHandling', 'GKKicking', 'GKPositioning', 'GKReflexes']

In [5]:
data = data[needed_columns]

In [6]:
data.head()

ID            Name  Age  Height  Weight  Overall  Potential  Growth  \
0  158023        L. Messi   35     169      67       91         91       0   
1  165153      K. Benzema   34     185      81       91         91       0   
2  188545  R. Lewandowski   33     185      81       91         91       0   
3  192985    K. De Bruyne   31     181      70       91         91       0   
4  231747       K. Mbappé   23     182      73       91         95       4   

   TotalStats  BaseStats  ... Penalties Composure  Marking  StandingTackle  \
0        2190        452  ...        75        96       20              35   
1        2147        455  ...        84        90       43              24   
2        2205        458  ...        90        88       35              42   
3        2303        483  ...        83        89       68              65   
4        2177        470  ...        80        88       26              34   

   SlidingTackle  GKDiving  GKHandling  GKKicking GKPositioning  GKReflexes  
0             24         6          11         15            14           8  
1             18        13          11          5             5           7  
2             19        15           6         12             8          10  
3             53        15          13          5            10          13  
4             32        13           5          7            11           6  

[5 rows x 65 columns]

### 1. Handling Missing values if any

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18539 entries, 0 to 18538
Data columns (total 65 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 18539 non-null  int64  
 1   Name               18539 non-null  object 
 2   Age                18539 non-null  int64  
 3   Height             18539 non-null  int64  
 4   Weight             18539 non-null  int64  
 5   Overall            18539 non-null  int64  
 6   Potential          18539 non-null  int64  
 7   Growth             18539 non-null  int64  
 8   TotalStats         18539 non-null  int64  
 9   BaseStats          18539 non-null  int64  
 10  BestPosition       18539 non-null  object 
 11  Club               18539 non-null  object 
 12  ValueEUR           18539 non-null  int64  
 13  WageEUR            18539 non-null  int64  
 14  ReleaseClause      18539 non-null  int64  
 15  ContractUntil      18447 non-null  float64
 16  ClubJoined         185

In [8]:
data.columns

Index(['ID', 'Name', 'Age', 'Height', 'Weight', 'Overall', 'Potential',
       'Growth', 'TotalStats', 'BaseStats', 'BestPosition', 'Club', 'ValueEUR',
       'WageEUR', 'ReleaseClause', 'ContractUntil', 'ClubJoined', 'OnLoad',
       'PreferredFoot', 'IntReputation', 'WeakFoot', 'Nationality',
       'SkillMoves', 'AttackingWorkRate', 'DefensiveWorkRate', 'PaceTotal',
       'ShootingTotal', 'PassingTotal', 'DribblingTotal', 'DefendingTotal',
       'PhysicalityTotal', 'Crossing', 'Finishing', 'HeadingAccuracy',
       'ShortPassing', 'Volleys', 'Dribbling', 'Curve', 'FKAccuracy',
       'LongPassing', 'BallControl', 'Acceleration', 'SprintSpeed', 'Agility',
       'Reactions', 'Balance', 'ShotPower', 'Jumping', 'Stamina', 'Strength',
       'LongShots', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Marking', 'StandingTackle', 'SlidingTackle',
       'GKDiving', 'GKHandling', 'GKKicking', 'GKPositioning', 'GKReflexes'],
      dtype='object')

Q. Function to iterate over all coumns and find which columns have null values and display their name

In [10]:
for i in data.columns:
  if data[i].isnull().sum()>0:
    print(i)

ContractUntil


In [13]:
# Break down of function
#data['Growth'].isnull()
#data['Growth'].isnull().sum()

0

In [14]:
# We will print more info sum of null values
for i in data.columns:
  if data[i].isnull().sum()>0:
    print(i, data[i].isnull().sum())

ContractUntil 92


Q Now lets see what percentage of data is this null value?

In [16]:
# Grabbing the no of rows in column
data.shape, data.shape[0]

((18539, 65), 18539)

In [17]:
# I need to write an equn -> (No of null values/Total values in that column)
(data['ContractUntil'].isnull().sum()/data.shape[0])*100

0.49625114623226707

**Intutiton on Missing Values:-** Now my missing values for the 'ContractUntil' is around 0.496% of data.

Thus this value is negligible and ***we can remove these rows*** without much affecting the training of model

In [20]:
#Step 1: We isolate the row where null values are present
data[data['ContractUntil'].isnull()]

ID         Name  Age  Height  Weight  Overall  Potential  Growth  \
438    204120    T. Vaclík   33     188      85       80         80       0   
447    140233     G. Ochoa   36     185      78       80         80       0   
548    156433  A. Talavera   39     186      85       79         79       0   
831    195086    D. Lovren   32     188      84       78         78       0   
865    205211    M. Borjan   34     196      84       77         77       0   
...       ...          ...  ...     ...     ...      ...        ...     ...   
15193  268986    A. Suhail   23     182      73       60         68       8   
16404  270843    D. Djamas   18     182      73       58         74      16   
16424  267767    S. Kittos   20     186      75       58         68      10   
17308  140814  S. Georgiou   17     183      77       55         70      15   
18043  225581     M. Gould   28     189      77       52         55       3   

       TotalStats  BaseStats  ... Penalties Composure  Marking  \
438          1309        432  ...        13        63       20   
447          1265        440  ...        16        60       21   
548          1345        446  ...        44        60       26   
831          1714        371  ...        52        69       76   
865          1360        423  ...        34        60       21   
...           ...        ...  ...       ...       ...      ...   
15193        1581        336  ...        37        49       55   
16404        1405        318  ...        30        45       45   
16424         891        314  ...        16        35       11   
17308        1461        319  ...        40        47       36   
18043         980        298  ...        27        22       15   

       StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
438                18             18        78          75         74   
447                10             12        82          73         70   
548                13             14        80          82         68   
831                82             83        14          14         13   
865                23             21        78          74         75   
...               ...            ...       ...         ...        ...   
15193              61             59        10           8          6   
16404              57             55         9          10          7   
16424              13             14        62          59         55   
17308              36             38        16           8         14   
18043              11             12        57          47         44   

      GKPositioning  GKReflexes  
438              82          83  
447              81          84  
548              78          80  
831              10           7  
865              74          78  
...             ...         ...  
15193            14           5  
16404             7          10  
16424            54          60  
17308             7          13  
18043            48          56  

[92 rows x 65 columns]

In [21]:
#step 2 We grab the index values of rows we want to drop
data[data['ContractUntil'].isnull()].index

Index([  438,   447,   548,   831,   865,   896,   931,  1038,  1086,  1188,
        1232,  1233,  1340,  1384,  1401,  1541,  1576,  1589,  1638,  1657,
        1901,  2107,  2255,  2543,  2952,  3103,  3203,  3248,  3273,  3300,
        3304,  3336,  3505,  3579,  3733,  3951,  4044,  4059,  4130,  4160,
        4176,  4522,  4592,  4810,  4961,  5537,  5599,  5831,  6022,  6227,
        6314,  6460,  6738,  6924,  6944,  7036,  7065,  7864,  8105,  8254,
        8312,  8624,  8770,  8975,  9282,  9453,  9639,  9901,  9903,  9943,
       10006, 10151, 10416, 10424, 10504, 10707, 11383, 11499, 11847, 12191,
       12351, 12939, 12989, 13761, 14575, 14591, 14901, 15193, 16404, 16424,
       17308, 18043],
      dtype='int64')

In [22]:
#Step 3 We drop the rows and use the inplace to save data
data.drop(data[data['ContractUntil'].isnull()].index, axis=0, inplace=True)

In [23]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18447 entries, 0 to 18538
Data columns (total 65 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 18447 non-null  int64  
 1   Name               18447 non-null  object 
 2   Age                18447 non-null  int64  
 3   Height             18447 non-null  int64  
 4   Weight             18447 non-null  int64  
 5   Overall            18447 non-null  int64  
 6   Potential          18447 non-null  int64  
 7   Growth             18447 non-null  int64  
 8   TotalStats         18447 non-null  int64  
 9   BaseStats          18447 non-null  int64  
 10  BestPosition       18447 non-null  object 
 11  Club               18447 non-null  object 
 12  ValueEUR           18447 non-null  int64  
 13  WageEUR            18447 non-null  int64  
 14  ReleaseClause      18447 non-null  int64  
 15  ContractUntil      18447 non-null  float64
 16  ClubJoined         18447 no

### 2. Handling the categorical columns

In [24]:
data['AttackingWorkRate'].dtype

dtype('O')

In [26]:
#Identifying the categorical columns

for i in data.columns:
  if data[i].dtype == 'object':
    print(i)

Name
BestPosition
Club
PreferredFoot
Nationality
AttackingWorkRate
DefensiveWorkRate


In [27]:
data['BestPosition']

0        CAM
1         CF
2         ST
3         CM
4         ST
        ... 
18534    CAM
18535    CDM
18536     RM
18537     CB
18538     LB
Name: BestPosition, Length: 18447, dtype: object

In [28]:
data['BestPosition'].unique()

array(['CAM', 'CF', 'ST', 'CM', 'RW', 'GK', 'CB', 'LW', 'CDM', 'LM', 'LB',
       'RM', 'RB', 'LWB', 'RWB'], dtype=object)

Identifying the unique classes in categorical columns

In [30]:
#Identifying the unique classes in categorical columns
for i in data.columns:
  if data[i].dtype == 'object':
    print(i, data[i].unique())

Name ['L. Messi' 'K. Benzema' 'R. Lewandowski' ... 'L. Mullan' 'D. McCallion'
 'N. Rabha']
BestPosition ['CAM' 'CF' 'ST' 'CM' 'RW' 'GK' 'CB' 'LW' 'CDM' 'LM' 'LB' 'RM' 'RB' 'LWB'
 'RWB']
Club ['Paris Saint-Germain' 'Real Madrid CF' 'FC Barcelona' 'Manchester City'
 'Liverpool' 'FC Bayern München' 'Manchester United' 'Tottenham Hotspur'
 'Atlético de Madrid' 'Chelsea' 'AC Milan' 'Lazio' 'Inter' 'Roma'
 'Juventus' 'Eintracht Frankfurt' 'Villarreal CF' 'RB Leipzig'
 'Borussia Dortmund' 'Real Betis Balompié' 'RC Celta de Vigo' 'Sevilla FC'
 'Leicester City' 'Bayer 04 Leverkusen' 'Borussia Mönchengladbach'
 'Galatasaray SK' 'Club Nacional de Football' 'Ajax'
 'U.S. Sassuolo Calcio' 'Arsenal' 'Aston Villa' 'AS Monaco' 'Toronto FC'
 'VfL Wolfsburg' 'Athletic Club de Bilbao' 'Newcastle United'
 'Los Angeles FC' 'AFC Richmond' 'West Ham United' 'Real Sociedad'
 'TSG Hoffenheim' 'Wolverhampton Wanderers' 'Palmeiras' 'Napoli' 'PSV'
 'Atalanta' 'OGC Nice' 'SL Benfica' 'Getafe CF' 'Al Nassr' 'Valenc

Identifying the unique classes in categorical columns

In [31]:
#Identifying the number of unique classes in categorical columns
for i in data.columns:
  if data[i].dtype == 'object':
    print(i, data[i].nunique())

Name 17448
BestPosition 15
Club 678
PreferredFoot 2
Nationality 159
AttackingWorkRate 3
DefensiveWorkRate 3


We will converth this from catagorical to numerical vector

In [32]:
from sklearn.preprocessing import LabelEncoder

In [34]:
le = LabelEncoder()

# We will do ordinal encoding ->

In [35]:
le.fit_transform(data['PreferredFoot'])

array([0, 1, 1, ..., 1, 1, 0])

In [36]:
le.fit_transform(data['AttackingWorkRate'])

array([1, 2, 0, ..., 0, 2, 2])

In [38]:
dd = le.fit_transform(data['BestPosition'])

In [40]:
dd.max(), dd.min()

(14, 0)

Lets implement LabelEncoder fit_transform

In [41]:
le = LabelEncoder()

data['PreferredFoot'] = le.fit_transform(data['PreferredFoot'])
data['AttackingWorkRate'] = le.fit_transform(data['AttackingWorkRate'])
data['DefensiveWorkRate'] = le.fit_transform(data['DefensiveWorkRate'])
data['Club'] = le.fit_transform(data['Club'])

#### Now we need to deal with our output parameter which is nothing but the 'BestPosition'

In [42]:
data['BestPosition'].unique(), data['BestPosition'].nunique()

array(['CAM', 'CF', 'ST', 'CM', 'RW', 'GK', 'CB', 'LW', 'CDM', 'LM', 'LB',
       'RM', 'RB', 'LWB', 'RWB'], dtype=object)

In [44]:
# We will use the domain knowledge to reduce our BestPosition parameters and merge common ones

merge_pos = {'LWB': 'LW', 'RWB': 'RW', 'ST': 'CF', 'CAM': 'CM', 'CDM': 'CM'}

In [46]:
data = data.replace({'BestPosition': merge_pos})

In [47]:
data['BestPosition'].unique(), data['BestPosition'].nunique()

(array(['CM', 'CF', 'RW', 'GK', 'CB', 'LW', 'LM', 'LB', 'RM', 'RB'],
       dtype=object),
 10)

In [45]:
position_map = {'CF': 0, 'CM': 1, 'RW': 2, 'GK': 3, 'CB': 4, 'LW': 5, 'LM': 6, 'LB': 7, 'RM':  8, 'RB': 9}

In [50]:
data = data.replace({'BestPosition': position_map})

In [51]:
data['BestPosition'].unique(), data['BestPosition'].nunique()

(array([1, 0, 2, 3, 4, 5, 6, 7, 8, 9]), 10)

In [52]:
data.head()

ID            Name  Age  Height  Weight  Overall  Potential  Growth  \
0  158023        L. Messi   35     169      67       91         91       0   
1  165153      K. Benzema   34     185      81       91         91       0   
2  188545  R. Lewandowski   33     185      81       91         91       0   
3  192985    K. De Bruyne   31     181      70       91         91       0   
4  231747       K. Mbappé   23     182      73       91         95       4   

   TotalStats  BaseStats  ...  Penalties  Composure  Marking  StandingTackle  \
0        2190        452  ...         75         96       20              35   
1        2147        455  ...         84         90       43              24   
2        2205        458  ...         90         88       35              42   
3        2303        483  ...         83         89       68              65   
4        2177        470  ...         80         88       26              34   

   SlidingTackle  GKDiving  GKHandling  GKKicking  GKPositioning  GKReflexes  
0             24         6          11         15             14           8  
1             18        13          11          5              5           7  
2             19        15           6         12              8          10  
3             53        15          13          5             10          13  
4             32        13           5          7             11           6  

[5 rows x 65 columns]

In [53]:
data.columns

Index(['ID', 'Name', 'Age', 'Height', 'Weight', 'Overall', 'Potential',
       'Growth', 'TotalStats', 'BaseStats', 'BestPosition', 'Club', 'ValueEUR',
       'WageEUR', 'ReleaseClause', 'ContractUntil', 'ClubJoined', 'OnLoad',
       'PreferredFoot', 'IntReputation', 'WeakFoot', 'Nationality',
       'SkillMoves', 'AttackingWorkRate', 'DefensiveWorkRate', 'PaceTotal',
       'ShootingTotal', 'PassingTotal', 'DribblingTotal', 'DefendingTotal',
       'PhysicalityTotal', 'Crossing', 'Finishing', 'HeadingAccuracy',
       'ShortPassing', 'Volleys', 'Dribbling', 'Curve', 'FKAccuracy',
       'LongPassing', 'BallControl', 'Acceleration', 'SprintSpeed', 'Agility',
       'Reactions', 'Balance', 'ShotPower', 'Jumping', 'Stamina', 'Strength',
       'LongShots', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Marking', 'StandingTackle', 'SlidingTackle',
       'GKDiving', 'GKHandling', 'GKKicking', 'GKPositioning', 'GKReflexes'],
      dtype='object')

### Finally obtaining clean data

In [54]:
data.head()

ID            Name  Age  Height  Weight  Overall  Potential  Growth  \
0  158023        L. Messi   35     169      67       91         91       0   
1  165153      K. Benzema   34     185      81       91         91       0   
2  188545  R. Lewandowski   33     185      81       91         91       0   
3  192985    K. De Bruyne   31     181      70       91         91       0   
4  231747       K. Mbappé   23     182      73       91         95       4   

   TotalStats  BaseStats  ...  Penalties  Composure  Marking  StandingTackle  \
0        2190        452  ...         75         96       20              35   
1        2147        455  ...         84         90       43              24   
2        2205        458  ...         90         88       35              42   
3        2303        483  ...         83         89       68              65   
4        2177        470  ...         80         88       26              34   

   SlidingTackle  GKDiving  GKHandling  GKKicking  GKPositioning  GKReflexes  
0             24         6          11         15             14           8  
1             18        13          11          5              5           7  
2             19        15           6         12              8          10  
3             53        15          13          5             10          13  
4             32        13           5          7             11           6  

[5 rows x 65 columns]

In [55]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18447 entries, 0 to 18538
Data columns (total 65 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 18447 non-null  int64  
 1   Name               18447 non-null  object 
 2   Age                18447 non-null  int64  
 3   Height             18447 non-null  int64  
 4   Weight             18447 non-null  int64  
 5   Overall            18447 non-null  int64  
 6   Potential          18447 non-null  int64  
 7   Growth             18447 non-null  int64  
 8   TotalStats         18447 non-null  int64  
 9   BaseStats          18447 non-null  int64  
 10  BestPosition       18447 non-null  int64  
 11  Club               18447 non-null  int64  
 12  ValueEUR           18447 non-null  int64  
 13  WageEUR            18447 non-null  int64  
 14  ReleaseClause      18447 non-null  int64  
 15  ContractUntil      18447 non-null  float64
 16  ClubJoined         18447 no

Save this dataset here ...


In [56]:
data.to_csv('fifa2023_cleaned.csv', index = False)